File: internal/api/setup_readiness.go
package api

import (
	"errors"
	"net/http"
	"sort"
	"time"

	"github.com/seanbao/mnemonas/internal/auth"
	"github.com/seanbao/mnemonas/internal/backup"
	"github.com/seanbao/mnemonas/internal/config"
)

const (
	setupLifecyclePending   = "pending"
	setupLifecycleDeferred  = "deferred"
	setupLifecycleCompleted = "completed"

	setupOverallReady          = "ready"
	setupOverallActionRequired = "action_required"
	setupOverallUnavailable    = "unavailable"

	setupRequirementRequired    = "required"
	setupRequirementRecommended = "recommended"

	setupCheckComplete      = "complete"
	setupCheckIncomplete    = "incomplete"
	setupCheckUnavailable   = "unavailable"
	setupCheckNotApplicable = "not_applicable"
)

type setupReadinessCount struct {
	Completed int `json:"completed"`
	Total     int `json:"total"`
}

type setupReadinessCheck struct {
	ID          string `json:"id"`
	Requirement string `json:"requirement"`
	Status      string `json:"status"`
	Deferrable  bool   `json:"deferrable"`
	Title       string `json:"title"`
	Message     string `json:"message"`
	Action      string `json:"action"`
}

type setupReadinessSummary struct {
	AuthEnabled                      bool       `json:"auth_enabled"`
	ActiveAdminCount                 int        `json:"active_admin_count"`
	PasswordChangeRequiredAdminCount int        `json:"password_change_required_admin_count"`
	InitialPasswordFile              string     `json:"initial_password_file"`
	EnabledBackupJobCount            int        `json:"enabled_backup_job_count"`
	LatestBackupSuccessAt            *time.Time `json:"latest_backup_success_at,omitempty"`
	LatestRestoreVerificationAt      *time.Time `json:"latest_restore_verification_at,omitempty"`
	SecurityStatus                   string     `json:"security_status"`
	SecurityBlockingCheckIDs         []string   `json:"security_blocking_check_ids"`
}

type setupReadinessData struct {
	Lifecycle     string                `json:"lifecycle"`
	Prompt        bool                  `json:"prompt"`
	GeneratedAt   time.Time             `json:"generated_at"`
	CompletedAt   *time.Time            `json:"completed_at,omitempty"`
	DeferredUntil *time.Time            `json:"deferred_until,omitempty"`
	OverallStatus string                `json:"overall_status"`
	CanComplete   bool                  `json:"can_complete"`
	CanDefer      bool                  `json:"can_defer"`
	Required      setupReadinessCount   `json:"required"`
	Recommended   setupReadinessCount   `json:"recommended"`
	Checks        []setupReadinessCheck `json:"checks"`
	Summary       setupReadinessSummary `json:"summary"`
}

type setupReadinessInputs struct {
	LifecycleAvailable bool
	Lifecycle          config.SetupLifecycleState
	AuthEnabled        bool
	UsersAvailable     bool
	Users              []*auth.User
	Backup             backup.ReadinessSnapshot
	SecurityAvailable  bool
	Security           securityCheckResponse
}

type deferSetupRequest struct {
	RemindInDays int `json:"remind_in_days"`
}

func setupCheck(id, requirement, status string, deferrable bool, title, message, action string) setupReadinessCheck {
	return setupReadinessCheck{
		ID:          id,
		Requirement: requirement,
		Status:      status,
		Deferrable:  deferrable,
		Title:       title,
		Message:     message,
		Action:      action,
	}
}

func evaluateSetupReadiness(inputs setupReadinessInputs, now time.Time) setupReadinessData {
	now = now.UTC()
	activeAdmins := 0
	passwordChangeRequiredAdmins := 0
	if inputs.UsersAvailable {
		for _, user := range inputs.Users {
			if user == nil || user.Disabled || user.Role != auth.RoleAdmin {
				continue
			}
			activeAdmins++
			if user.MustChangePassword {
				passwordChangeRequiredAdmins++
			}
		}
	}

	checks := make([]setupReadinessCheck, 0, 10)
	switch {
	case !inputs.AuthEnabled:
		checks = append(checks, setupCheck("admin_access", setupRequirementRequired, setupCheckIncomplete, false, "启用管理员登录", "管理界面尚未启用账号登录。", "review_security"))
	case !inputs.UsersAvailable:
		checks = append(checks, setupCheck("admin_access", setupRequirementRequired, setupCheckUnavailable, false, "验证管理员访问", "当前无法读取管理员账号状态。", "manage_users"))
	case activeAdmins == 0:
		checks = append(checks, setupCheck("admin_access", setupRequirementRequired, setupCheckIncomplete, false, "保留可用管理员", "没有启用中的管理员账号。", "manage_users"))
	default:
		checks = append(checks, setupCheck("admin_access", setupRequirementRequired, setupCheckComplete, false, "管理员访问可用", "至少有一个启用中的管理员账号。", "manage_users"))
	}

	switch {
	case !inputs.AuthEnabled:
		checks = append(checks, setupCheck("bootstrap_credential", setupRequirementRequired, setupCheckIncomplete, false, "保护初始凭据", "启用管理员登录后才能验证初始凭据是否已更换。", "review_security"))
	case !inputs.UsersAvailable:
		checks = append(checks, setupCheck("bootstrap_credential", setupRequirementRequired, setupCheckUnavailable, false, "验证初始凭据", "当前无法读取管理员密码状态。", "change_password"))
	case activeAdmins == 0:
		checks = append(checks, setupCheck("bootstrap_credential", setupRequirementRequired, setupCheckIncomplete, false, "保护初始凭据", "需要先创建一个可用管理员账号。", "manage_users"))
	case passwordChangeRequiredAdmins > 0:
		checks = append(checks, setupCheck("bootstrap_credential", setupRequirementRequired, setupCheckIncomplete, false, "更换初始密码", "仍有管理员使用初始化流程生成的密码。", "change_password"))
	default:
		checks = append(checks, setupCheck("bootstrap_credential", setupRequirementRequired, setupCheckComplete, false, "初始密码已更换", "启用中的管理员均已完成密码更换。", "change_password"))
	}

	initialPasswordStatus := setupCheckUnavailable
	initialPasswordSummary := "unavailable"
	securityBlockingIDs := make([]string, 0)
	securityWarning := false
	if inputs.SecurityAvailable {
		for _, check := range inputs.Security.Checks {
			if check.Status == securityCheckBlock {
				securityBlockingIDs = append(securityBlockingIDs, check.ID)
			}
			if check.Status == securityCheckWarning {
				securityWarning = true
			}
			if check.ID != "initial_password_file" {
				continue
			}
			switch check.Status {
			case securityCheckPass:
				initialPasswordStatus = setupCheckComplete
				initialPasswordSummary = "missing"
			case securityCheckBlock:
				initialPasswordStatus = setupCheckIncomplete
				initialPasswordSummary = "present"
			default:
				initialPasswordStatus = setupCheckUnavailable
				initialPasswordSummary = "unavailable"
			}
		}
	}
	sort.Strings(securityBlockingIDs)
	securityBlockingIDs = compactSortedStrings(securityBlockingIDs)
	initialPasswordMessage := "无法验证服务器上的初始密码文件。"
	if initialPasswordStatus == setupCheckComplete {
		initialPasswordMessage = "服务器上没有遗留初始密码文件。"
	} else if initialPasswordStatus == setupCheckIncomplete {
		initialPasswordMessage = "服务器上仍存在初始密码文件或其路径不安全。"
	}
	checks = append(checks, setupCheck("initial_password_file", setupRequirementRequired, initialPasswordStatus, false, "清理初始密码文件", initialPasswordMessage, "change_password"))

	baselineBlockingIDs := make([]string, 0, len(securityBlockingIDs))
	for _, id := range securityBlockingIDs {
		if id != "initial_password_file" && id != "admin_accounts" {
			baselineBlockingIDs = append(baselineBlockingIDs, id)
		}
	}
	securityBaselineStatus := setupCheckComplete
	securityBaselineMessage := "安全基线没有阻断项。"
	if !inputs.SecurityAvailable {
		securityBaselineStatus = setupCheckUnavailable
		securityBaselineMessage = "当前无法完成安全基线检查。"
	} else if len(baselineBlockingIDs) > 0 {
		securityBaselineStatus = setupCheckIncomplete
		securityBaselineMessage = "安全自检仍有必须处理的阻断项。"
	}
	checks = append(checks, setupCheck("security_baseline", setupRequirementRequired, securityBaselineStatus, false, "满足安全基线", securityBaselineMessage, "review_security"))

	backupJobStatus := setupCheckIncomplete
	backupJobMessage := "尚未添加启用中的备份任务。"
	if !inputs.Backup.Available {
		backupJobStatus = setupCheckUnavailable
		backupJobMessage = "当前无法读取备份任务状态。"
	} else if inputs.Backup.EnabledJobCount > 0 {
		backupJobStatus = setupCheckComplete
		backupJobMessage = "已配置启用中的独立备份任务。"
	}
	checks = append(checks, setupCheck("backup_job", setupRequirementRequired, backupJobStatus, true, "添加独立备份", backupJobMessage, "create_backup"))

	backupSuccessStatus := setupCheckIncomplete
	backupSuccessMessage := "尚无当前有效的成功备份。"
	if !inputs.Backup.Available {
		backupSuccessStatus = setupCheckUnavailable
		backupSuccessMessage = "当前无法读取最近备份结果。"
	} else if inputs.Backup.HasCurrentHealthyBackup {
		backupSuccessStatus = setupCheckComplete
		backupSuccessMessage = "已有当前有效的成功备份。"
	}
	checks = append(checks, setupCheck("backup_success", setupRequirementRequired, backupSuccessStatus, true, "完成首次备份", backupSuccessMessage, "run_backup"))

	adminRedundancyStatus := setupCheckIncomplete
	adminRedundancyMessage := "建议再准备一个启用中的管理员账号。"
	if !inputs.UsersAvailable {
		adminRedundancyStatus = setupCheckUnavailable
		adminRedundancyMessage = "当前无法验证备用管理员。"
	} else if activeAdmins >= 2 {
		adminRedundancyStatus = setupCheckComplete
		adminRedundancyMessage = "已有备用管理员账号。"
	}
	checks = append(checks, setupCheck("admin_redundancy", setupRequirementRecommended, adminRedundancyStatus, false, "准备备用管理员", adminRedundancyMessage, "manage_users"))

	backupScheduleStatus := setupCheckIncomplete
	backupScheduleMessage := "建议为备份任务启用自动计划。"
	if !inputs.Backup.Available {
		backupScheduleStatus = setupCheckUnavailable
		backupScheduleMessage = "当前无法验证备份计划。"
	} else if inputs.Backup.EnabledScheduledJobCount > 0 {
		backupScheduleStatus = setupCheckComplete
		backupScheduleMessage = "至少有一个启用中的自动备份计划。"
	}
	checks = append(checks, setupCheck("backup_schedule", setupRequirementRecommended, backupScheduleStatus, false, "启用自动备份", backupScheduleMessage, "create_backup"))

	restoreStatus := setupCheckIncomplete
	restoreMessage := "建议执行一次恢复演练并保持验证结果有效。"
	if !inputs.Backup.Available {
		restoreStatus = setupCheckUnavailable
		restoreMessage = "当前无法读取恢复验证状态。"
	} else if inputs.Backup.HasCurrentRestoreEvidence {
		restoreStatus = setupCheckComplete
		restoreMessage = "已有当前有效的恢复验证记录。"
	}
	checks = append(checks, setupCheck("restore_verification", setupRequirementRecommended, restoreStatus, false, "验证恢复能力", restoreMessage, "run_restore_drill"))

	securityRecommendationStatus := setupCheckIncomplete
	securityRecommendationMessage := "安全自检仍有建议处理的项目。"
	if !inputs.SecurityAvailable {
		securityRecommendationStatus = setupCheckUnavailable
		securityRecommendationMessage = "当前无法读取安全建议。"
	} else if !securityWarning && len(securityBlockingIDs) == 0 {
		securityRecommendationStatus = setupCheckComplete
		securityRecommendationMessage = "安全自检全部通过。"
	}
	checks = append(checks, setupCheck("security_recommendations", setupRequirementRecommended, securityRecommendationStatus, false, "处理安全建议", securityRecommendationMessage, "review_security"))

	required := countSetupChecks(checks, setupRequirementRequired)
	recommended := countSetupChecks(checks, setupRequirementRecommended)
	requiredUnavailable := false
	nonDeferrableReady := true
	deferrableIncomplete := false
	deferrableUnavailable := false
	for _, check := range checks {
		if check.Requirement != setupRequirementRequired {
			continue
		}
		if check.Status == setupCheckUnavailable {
			requiredUnavailable = true
		}
		if check.Deferrable {
			if check.Status == setupCheckUnavailable {
				deferrableUnavailable = true
			}
			if check.Status == setupCheckIncomplete {
				deferrableIncomplete = true
			}
			continue
		}
		if check.Status != setupCheckComplete && check.Status != setupCheckNotApplicable {
			nonDeferrableReady = false
		}
	}

	overallStatus := setupOverallActionRequired
	canComplete := required.Completed == required.Total
	if !inputs.LifecycleAvailable || requiredUnavailable {
		overallStatus = setupOverallUnavailable
		canComplete = false
	} else if canComplete {
		overallStatus = setupOverallReady
	}
	canDefer := inputs.LifecycleAvailable && nonDeferrableReady && deferrableIncomplete && !deferrableUnavailable

	lifecycle := setupLifecyclePending
	prompt := inputs.LifecycleAvailable
	var completedAt *time.Time
	var deferredUntil *time.Time
	if inputs.LifecycleAvailable {
		completedAt = cloneSetupTime(inputs.Lifecycle.CompletedAt)
		deferredUntil = cloneSetupTime(inputs.Lifecycle.DeferredUntil)
		switch {
		case completedAt != nil:
			lifecycle = setupLifecycleCompleted
			prompt = false
			canComplete = false
			canDefer = false
		case deferredUntil != nil && now.Before(*deferredUntil) && nonDeferrableReady && deferrableIncomplete && !requiredUnavailable:
			lifecycle = setupLifecycleDeferred
			prompt = false
		default:
			lifecycle = setupLifecyclePending
			prompt = true
		}
	}

	securityStatus := "unavailable"
	if inputs.SecurityAvailable {
		securityStatus = string(inputs.Security.Status)
	}
	return setupReadinessData{
		Lifecycle:     lifecycle,
		Prompt:        prompt,
		GeneratedAt:   now,
		CompletedAt:   completedAt,
		DeferredUntil: deferredUntil,
		OverallStatus: overallStatus,
		CanComplete:   canComplete,
		CanDefer:      canDefer,
		Required:      required,
		Recommended:   recommended,
		Checks:        checks,
		Summary: setupReadinessSummary{
			AuthEnabled:                      inputs.AuthEnabled,
			ActiveAdminCount:                 activeAdmins,
			PasswordChangeRequiredAdminCount: passwordChangeRequiredAdmins,
			InitialPasswordFile:              initialPasswordSummary,
			EnabledBackupJobCount:            inputs.Backup.EnabledJobCount,
			LatestBackupSuccessAt:            cloneSetupTime(inputs.Backup.LastSuccessfulBackupAt),
			LatestRestoreVerificationAt:      cloneSetupTime(inputs.Backup.LastValidRestoreEvidenceAt),
			SecurityStatus:                   securityStatus,
			SecurityBlockingCheckIDs:         securityBlockingIDs,
		},
	}
}

func countSetupChecks(checks []setupReadinessCheck, requirement string) setupReadinessCount {
	count := setupReadinessCount{}
	for _, check := range checks {
		if check.Requirement != requirement {
			continue
		}
		count.Total++
		if check.Status == setupCheckComplete || check.Status == setupCheckNotApplicable {
			count.Completed++
		}
	}
	return count
}

func cloneSetupTime(value *time.Time) *time.Time {
	if value == nil {
		return nil
	}
	cloned := value.UTC()
	return &cloned
}

func compactSortedStrings(values []string) []string {
	if len(values) < 2 {
		return values
	}
	write := 1
	for read := 1; read < len(values); read++ {
		if values[read] == values[write-1] {
			continue
		}
		values[write] = values[read]
		write++
	}
	return values[:write]
}

func (s *Server) buildSetupReadiness(r *http.Request, now time.Time) setupReadinessData {
	inputs := setupReadinessInputs{
		AuthEnabled:    s.authEnabled,
		UsersAvailable: s.userStore != nil,
	}
	if s.backupManager != nil {
		inputs.Backup = s.backupManager.ReadinessSnapshotContext(r.Context())
	}
	if s.userStore != nil {
		inputs.Users = s.userStore.List()
	}

	cfg := s.currentConfig()
	if cfg != nil {
		secrets, err := config.LoadSecrets(cfg.Storage.Root)
		if err == nil && secrets != nil {
			inputs.LifecycleAvailable = true
			inputs.Lifecycle = secrets.SetupLifecycle
		}
	}
	security, err := s.buildSecurityCheck(r)
	if err == nil && security != nil {
		inputs.SecurityAvailable = true
		inputs.Security = *security
	}
	return evaluateSetupReadiness(inputs, now)
}

func (s *Server) handleGetSetupReadiness(w http.ResponseWriter, r *http.Request) {
	setSetupReadResponseHeaders(w)
	readiness := s.buildSetupReadiness(r, apiTimeNow())
	NewAPIResponse(readiness).Write(w, http.StatusOK)
}

func setSetupReadResponseHeaders(w http.ResponseWriter) {
	header := w.Header()
	header.Set("Cache-Control", "private, no-store")
	appendVaryHeaderToken(header, "Cookie")
	appendVaryHeaderToken(header, "Authorization")
}

func withSetupReadResponseHeaders(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		setSetupReadResponseHeaders(w)
		next.ServeHTTP(w, r)
	})
}

func (s *Server) handleAcknowledgeSetup(w http.ResponseWriter, r *http.Request) {
	if !s.authEnabled {
		Forbidden(w, "setup completion requires administrator authentication")
		return
	}
	var req struct{}
	if err := decodeJSONBody(r, &req); err != nil {
		writeLimitedJSONBodyError(w, err, DefaultJSONRequestBodyLimit)
		return
	}

	s.setupMu.Lock()
	defer s.setupMu.Unlock()
	s.settingsMu.Lock()
	defer s.settingsMu.Unlock()
	now := apiTimeNow().UTC()
	readiness := s.buildSetupReadiness(r, now)
	if readiness.Lifecycle == setupLifecycleCompleted {
		NewAPIResponse(readiness).WithMessage("setup already completed").Write(w, http.StatusOK)
		return
	}
	if readiness.OverallStatus == setupOverallUnavailable {
		writeSetupReadinessError(w, http.StatusServiceUnavailable, "SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable", readiness, true)
		return
	}
	if !readiness.CanComplete {
		writeSetupReadinessError(w, http.StatusConflict, "SETUP_NOT_READY", "required setup checks are incomplete", readiness, false)
		return
	}

	cfg := s.currentConfig()
	if cfg == nil {
		NewAPIError("SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable").Write(w, http.StatusServiceUnavailable)
		return
	}
	if _, err := config.CompleteSetup(cfg.Storage.Root, now); err != nil {
		if errors.Is(err, config.ErrSecretsNotFound) {
			NewAPIError("SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable").Write(w, http.StatusServiceUnavailable)
			return
		}
		s.respondInternalError(w, "complete setup", err)
		return
	}
	updated := s.buildSetupReadiness(r, now)
	NewAPIResponse(updated).WithMessage("setup completed").Write(w, http.StatusOK)
}

func (s *Server) handleDeferSetup(w http.ResponseWriter, r *http.Request) {
	if !s.authEnabled {
		Forbidden(w, "setup deferral requires administrator authentication")
		return
	}
	var req deferSetupRequest
	if err := decodeJSONBody(r, &req); err != nil {
		writeLimitedJSONBodyError(w, err, DefaultJSONRequestBodyLimit)
		return
	}
	if req.RemindInDays < 1 || req.RemindInDays > 30 {
		BadRequest(w, "remind_in_days must be between 1 and 30")
		return
	}

	s.setupMu.Lock()
	defer s.setupMu.Unlock()
	s.settingsMu.Lock()
	defer s.settingsMu.Unlock()
	now := apiTimeNow().UTC()
	readiness := s.buildSetupReadiness(r, now)
	if readiness.Lifecycle == setupLifecycleCompleted {
		writeSetupReadinessError(w, http.StatusConflict, "SETUP_DEFER_FORBIDDEN", "setup cannot be deferred", readiness, false)
		return
	}
	if readiness.OverallStatus == setupOverallUnavailable {
		writeSetupReadinessError(w, http.StatusServiceUnavailable, "SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable", readiness, true)
		return
	}
	if !readiness.CanDefer {
		writeSetupReadinessError(w, http.StatusConflict, "SETUP_DEFER_FORBIDDEN", "setup cannot be deferred", readiness, false)
		return
	}

	cfg := s.currentConfig()
	if cfg == nil {
		NewAPIError("SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable").Write(w, http.StatusServiceUnavailable)
		return
	}
	until := now.AddDate(0, 0, req.RemindInDays)
	if _, err := config.DeferSetup(cfg.Storage.Root, now, until); err != nil {
		if errors.Is(err, config.ErrSecretsNotFound) {
			NewAPIError("SETUP_READINESS_UNAVAILABLE", "setup readiness is unavailable").Write(w, http.StatusServiceUnavailable)
			return
		}
		s.respondInternalError(w, "defer setup", err)
		return
	}
	updated := s.buildSetupReadiness(r, now)
	NewAPIResponse(updated).WithMessage("setup deferred").Write(w, http.StatusOK)
}

func writeSetupReadinessError(w http.ResponseWriter, status int, code, message string, readiness setupReadinessData, includeUnavailable bool) {
	ids := make([]string, 0)
	for _, check := range readiness.Checks {
		if check.Requirement != setupRequirementRequired || check.Status == setupCheckComplete || check.Status == setupCheckNotApplicable {
			continue
		}
		if !includeUnavailable && check.Status == setupCheckUnavailable {
			continue
		}
		ids = append(ids, check.ID)
	}
	sort.Strings(ids)
	NewAPIError(code, message).WithDetails(map[string]any{"required_check_ids": ids}).Write(w, status)
}

File: web/e2e/helpers/setup-readiness.ts
import type { Page, Route } from '@playwright/test'

type SetupReadinessAction =
  | 'change_password'
  | 'manage_users'
  | 'create_backup'
  | 'run_backup'
  | 'run_restore_drill'
  | 'review_security'

type SetupReadinessCheck = {
  id: string
  requirement: 'required' | 'recommended'
  status: 'complete' | 'incomplete' | 'unavailable' | 'not_applicable'
  deferrable: boolean
  title: string
  message: string
  action: SetupReadinessAction
}

type SetupReadinessSummary = {
  auth_enabled: boolean
  active_admin_count: number
  password_change_required_admin_count: number
  initial_password_file: 'missing' | 'present' | 'unavailable'
  enabled_backup_job_count: number
  latest_backup_success_at?: string
  latest_restore_verification_at?: string
  security_status: 'pass' | 'warning' | 'block' | 'unavailable'
  security_blocking_check_ids: string[]
}

export type SetupReadiness = {
  lifecycle: 'pending' | 'deferred' | 'completed'
  overall_status: 'ready' | 'action_required' | 'unavailable'
  prompt: boolean
  generated_at: string
  completed_at?: string
  deferred_until?: string
  can_complete: boolean
  can_defer: boolean
  required: { completed: number; total: number }
  recommended: { completed: number; total: number }
  checks: SetupReadinessCheck[]
  summary: SetupReadinessSummary
}

type DeferSetupRequest = {
  remind_in_days: 1 | 3 | 7 | 30
}

const setupReadinessRoutePattern = /\/api\/v1\/setup\/(?:readiness|acknowledge|defer)(?:\?.*)?$/
const generatedAt = '2026-07-13T01:02:03Z'

type SetupCheckOverrides = Partial<Omit<SetupReadinessCheck, 'id'>>

export type SetupReadinessFactoryOptions = {
  lifecycle?: SetupReadiness['lifecycle']
  checkOverrides?: Record<string, SetupCheckOverrides>
  summary?: Partial<SetupReadinessSummary>
  completedAt?: string
  deferredUntil?: string
  generatedAt?: string
}

export type CapturedSetupReadinessRequest = {
  method: string
  pathname: string
  body: unknown
}

export type SetupReadinessMockReply = {
  status: number
  readiness?: SetupReadiness
  message?: string
  error?: {
    code: string
    message: string
    details?: unknown
  }
}

export type SetupReadinessRouteState = {
  current: SetupReadiness
  requests: {
    readiness: CapturedSetupReadinessRequest[]
    acknowledge: CapturedSetupReadinessRequest[]
    defer: CapturedSetupReadinessRequest[]
  }
  setReadiness: (readiness: SetupReadiness) => void
}

type SetupReadinessMutationHandler = (
  request: CapturedSetupReadinessRequest,
  state: SetupReadinessRouteState,
) => SetupReadinessMockReply | Promise<SetupReadinessMockReply>

export type SetupReadinessRouteOptions = {
  initialReadiness: SetupReadiness
  onAcknowledge?: SetupReadinessMutationHandler
  onDefer?: SetupReadinessMutationHandler
}

function check(
  id: string,
  requirement: SetupReadinessCheck['requirement'],
  title: string,
  message: string,
  action: SetupReadinessAction,
  deferrable: boolean,
): SetupReadinessCheck {
  return {
    id,
    requirement,
    status: 'complete',
    deferrable,
    title,
    message,
    action,
  }
}

const defaultChecks: SetupReadinessCheck[] = [
  check('admin_access', 'required', '管理员访问可用', '至少有一个启用中的管理员账号。', 'manage_users', false),
  check('bootstrap_credential', 'required', '初始密码已更换', '启用中的管理员均已完成密码更换。', 'change_password', false),
  check('initial_password_file', 'required', '清理初始密码文件', '服务器上没有遗留初始密码文件。', 'change_password', false),
  check('security_baseline', 'required', '满足安全基线', '安全基线没有阻断项。', 'review_security', false),
  check('backup_job', 'required', '添加独立备份', '已配置启用中的独立备份任务。', 'create_backup', true),
  check('backup_success', 'required', '完成首次备份', '已有当前有效的成功备份。', 'run_backup', true),
  check('admin_redundancy', 'recommended', '准备备用管理员', '已有备用管理员账号。', 'manage_users', false),
  check('backup_schedule', 'recommended', '启用自动备份', '至少有一个启用中的自动备份计划。', 'create_backup', false),
  check('restore_verification', 'recommended', '验证恢复能力', '已有当前有效的恢复验证记录。', 'run_restore_drill', false),
  check('security_recommendations', 'recommended', '处理安全建议', '安全自检全部通过。', 'review_security', false),
]

const defaultSummary: SetupReadinessSummary = {
  auth_enabled: true,
  active_admin_count: 2,
  password_change_required_admin_count: 0,
  initial_password_file: 'missing',
  enabled_backup_job_count: 1,
  latest_backup_success_at: '2026-07-13T00:02:03Z',
  latest_restore_verification_at: '2026-07-13T00:32:03Z',
  security_status: 'pass',
  security_blocking_check_ids: [],
}

function isComplete(checkItem: SetupReadinessCheck): boolean {
  return checkItem.status === 'complete' || checkItem.status === 'not_applicable'
}

function deriveReadinessGates(checks: SetupReadinessCheck[]) {
  const requiredChecks = checks.filter(checkItem => checkItem.requirement === 'required')
  const recommendedChecks = checks.filter(checkItem => checkItem.requirement === 'recommended')
  const requiredUnavailable = requiredChecks.some(checkItem => checkItem.status === 'unavailable')
  const nonDeferrableReady = requiredChecks
    .filter(checkItem => !checkItem.deferrable)
    .every(isComplete)
  const deferrableIncomplete = requiredChecks
    .filter(checkItem => checkItem.deferrable)
    .some(checkItem => checkItem.status === 'incomplete')
  const deferrableUnavailable = requiredChecks
    .filter(checkItem => checkItem.deferrable)
    .some(checkItem => checkItem.status === 'unavailable')
  const canComplete = requiredChecks.every(isComplete)

  return {
    overallStatus: requiredUnavailable
      ? 'unavailable' as const
      : canComplete
        ? 'ready' as const
        : 'action_required' as const,
    canComplete,
    canDefer: nonDeferrableReady && deferrableIncomplete && !deferrableUnavailable,
    required: {
      completed: requiredChecks.filter(isComplete).length,
      total: requiredChecks.length,
    },
    recommended: {
      completed: recommendedChecks.filter(isComplete).length,
      total: recommendedChecks.length,
    },
  }
}

function cloneReadiness(readiness: SetupReadiness): SetupReadiness {
  return JSON.parse(JSON.stringify(readiness)) as SetupReadiness
}

export function createSetupReadiness(options: SetupReadinessFactoryOptions = {}): SetupReadiness {
  const checks = defaultChecks.map(checkItem => ({
    ...checkItem,
    ...options.checkOverrides?.[checkItem.id],
  }))
  const gates = deriveReadinessGates(checks)
  const lifecycle = options.lifecycle ?? 'pending'

  return {
    lifecycle,
    overall_status: gates.overallStatus,
    prompt: lifecycle === 'pending',
    generated_at: options.generatedAt ?? generatedAt,
    ...(options.completedAt ? { completed_at: options.completedAt } : {}),
    ...(options.deferredUntil ? { deferred_until: options.deferredUntil } : {}),
    can_complete: lifecycle === 'completed' ? false : gates.canComplete,
    can_defer: lifecycle === 'completed' ? false : gates.canDefer,
    required: gates.required,
    recommended: gates.recommended,
    checks,
    summary: {
      ...defaultSummary,
      ...options.summary,
    },
  }
}

export function completedSetupReadiness(readiness: SetupReadiness): SetupReadiness {
  return {
    ...cloneReadiness(readiness),
    lifecycle: 'completed',
    prompt: false,
    completed_at: readiness.generated_at,
    deferred_until: undefined,
    can_complete: false,
    can_defer: false,
  }
}

export function deferredSetupReadiness(readiness: SetupReadiness, remindInDays: number): SetupReadiness {
  const deferredUntil = new Date(readiness.generated_at)
  deferredUntil.setUTCDate(deferredUntil.getUTCDate() + remindInDays)
  return {
    ...cloneReadiness(readiness),
    lifecycle: 'deferred',
    prompt: false,
    completed_at: undefined,
    deferred_until: deferredUntil.toISOString(),
  }
}

function requestBody(route: Route): unknown {
  const rawBody = route.request().postData()
  if (!rawBody) {
    return null
  }
  try {
    return JSON.parse(rawBody) as unknown
  } catch {
    return rawBody
  }
}

function capturedRequest(route: Route): CapturedSetupReadinessRequest {
  const request = route.request()
  return {
    method: request.method(),
    pathname: new URL(request.url()).pathname,
    body: requestBody(route),
  }
}

async function fulfillSuccess(route: Route, readiness: SetupReadiness, message?: string) {
  await route.fulfill({
    status: 200,
    contentType: 'application/json',
    body: JSON.stringify({
      success: true,
      data: cloneReadiness(readiness),
      ...(message ? { message } : {}),
      timestamp: generatedAt,
    }),
  })
}

async function fulfillReply(
  route: Route,
  state: SetupReadinessRouteState,
  reply: SetupReadinessMockReply,
) {
  if (reply.status >= 200 && reply.status < 300 && reply.readiness) {
    state.setReadiness(reply.readiness)
    await fulfillSuccess(route, state.current, reply.message)
    return
  }

  const error = reply.error ?? {
    code: 'SETUP_MOCK_ERROR',
    message: 'setup readiness mock rejected the request',
  }
  await route.fulfill({
    status: reply.status,
    contentType: 'application/json',
    body: JSON.stringify({
      ...error,
      timestamp: generatedAt,
    }),
  })
}

export async function routeSetupReadiness(
  page: Page,
  options: SetupReadinessRouteOptions,
): Promise<SetupReadinessRouteState> {
  const state: SetupReadinessRouteState = {
    current: cloneReadiness(options.initialReadiness),
    requests: {
      readiness: [],
      acknowledge: [],
      defer: [],
    },
    setReadiness(readiness) {
      state.current = cloneReadiness(readiness)
    },
  }

  await page.route(setupReadinessRoutePattern, async (route) => {
    const request = capturedRequest(route)

    if (request.pathname.endsWith('/readiness') && request.method === 'GET') {
      state.requests.readiness.push(request)
      await fulfillSuccess(route, state.current)
      return
    }

    if (request.pathname.endsWith('/acknowledge') && request.method === 'POST') {
      state.requests.acknowledge.push(request)
      const reply = options.onAcknowledge
        ? await options.onAcknowledge(request, state)
        : {
            status: 200,
            readiness: completedSetupReadiness(state.current),
            message: 'setup completed',
          }
      await fulfillReply(route, state, reply)
      return
    }

    if (request.pathname.endsWith('/defer') && request.method === 'POST') {
      state.requests.defer.push(request)
      const body = request.body as Partial<DeferSetupRequest> | null
      const remindInDays = typeof body?.remind_in_days === 'number' ? body.remind_in_days : 7
      const reply = options.onDefer
        ? await options.onDefer(request, state)
        : {
            status: 200,
            readiness: deferredSetupReadiness(state.current, remindInDays),
            message: 'setup deferred',
          }
      await fulfillReply(route, state, reply)
      return
    }

    await route.fulfill({ status: 405, body: '' })
  })

  return state
}
